### ISSUE TITLE
SCryptUtil.check does not work with 8 bit r/p

### ISSUE THREAD

**Issue by statop** [wg/scrypt#5]
The following code in SCryptUtil.check only extracts 4 bits for r and p

```
        int N = (int) Math.pow(2, params >> 16 & 0xff);
        int r = params >> 8 & 0x0f;
        int p = params      & 0x0f;
```

**Comment by statop** [comment]
also, you don't need 16 bits for N, it will not ever need more that 6 bits, so you can divide the rest between r and p

**Comment by statop** [comment]
fixed code with different bit sizes (might need to change version number):

``` java

    /**
     * Hash the supplied plaintext password and generate output in the format described
     * in {@link SCryptUtil}.
     *
     * @param passwd    Password.
     * @param salt      The salt to use
     * @param N         CPU cost parameter.
     * @param r         Memory cost parameter.
     * @param p         Parallelization parameter.
     *
     * @return The hashed password.
     */
    public static String scrypt(String passwd, byte[] salt, int N, int r, int p) {
        try {
            if (r > 8191) {
                throw new IllegalArgumentException("r must be less than 8192");
            }
            if (p > 8191) {
                throw new IllegalArgumentException("p must be less than 8192");
            }
            if (N > 1073741824) {
                throw new IllegalArgumentException("N must be less than 2^30 + 1");
            }
            byte[] derived = SCrypt.scrypt(passwd.getBytes("UTF-8"), salt, N, r, p, 128);

            String params = Integer.toString(log2(N) << 26 | r << 13 | p, 16);

            StringBuilder sb = new StringBuilder((salt.length + derived.length) * 2);
            sb.append("$s0$").append(params).append('$');
            sb.append(encodeToChar(salt, false)).append('$');
            sb.append(encodeToChar(derived, false));

            return sb.toString();
        } catch (UnsupportedEncodingException e) {
            throw new IllegalStateException("JVM doesn't support UTF-8?");
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("JVM doesn't support SHA1PRNG or HMAC_SHA256?");
        }
    }

    /**
     * Compare the supplied plaintext password to a hashed password.
     *
     * @param   passwd  Plaintext password.
     * @param   hashed  scrypt hashed password.
     *
     * @return true if passwd matches hashed value.
     */
    public static boolean check(String passwd, String hashed) {
        try {
            String[] parts = hashed.split("\\$");

            if (parts.length != 5 || !parts[1].equals("s0")) {
                throw new IllegalArgumentException("Invalid hashed value");
            }

            int params = Integer.parseInt(parts[2], 16);
            byte[] salt = decodeFast(parts[3].toCharArray());
            byte[] derived0 = decodeFast(parts[4].toCharArray());

            int N = (int) Math.pow(2, params >> 26);
            int r = params >> 13 & 0x1fff;
            int p = params       & 0x1fff;

            byte[] derived1 = SCrypt.scrypt(passwd.getBytes("UTF-8"), salt, N, r, p, 128);

            if (derived0.length != derived1.length) {
                return false;
            }

            int result = 0;
            for (int i = 0; i < derived0.length; i++) {
                result |= derived0[i] ^ derived1[i];
            }
            return result == 0;
        } catch (UnsupportedEncodingException e) {
            throw new IllegalStateException("JVM doesn't support UTF-8?");
        } catch (GeneralSecurityException e) {
            throw new IllegalStateException("JVM doesn't support SHA1PRNG or HMAC_SHA256?");
        }
    }
```

**Comment by wg** [comment]
Thank you! Right you are, and I've released version 1.3.3 which fixes the broken parameter bitmasking.
